h/services/annotation_metadata.py

```diff
@@ -0,0 +1,48 @@
+import os
+
+from sqlalchemy.dialects.postgresql import insert
+from sqlalchemy.orm import Session
+
+from h.models import Annotation, AnnotationMetadata, User
+from h.security import decrypt_jwe_dict
+
+
+class AnnotationMetadataService:
+    def __init__(self, db: Session):
+        self._db = db
+        self._secret = os.environ.get("JWE_SECRET_LMS")
+
+    def set_annotation_metadata_from_jwe(self, annotation: Annotation, jwe: str):
+        """
+        Set the decrypted `jwe` dict as the metadata for `annotation`.
+
+        It will create a new AnnotationMetadata row or update the existing one.
+        """
+        if not annotation.pk and not annotation.user_id:
+            # Temporarily add these to the annotation until
+            # h.task.annotations.fill_pk_and_user_id completes
+            annotation.pk = Annotation.pk_sequence.next_value()
+            annotation.user_id = (
+                self._db.query(User).filter_by(userid=annotation.userid).one().id
+            )
+            self._db.flush()
+
+        annotation_metadata = decrypt_jwe_dict(self._secret, jwe)
```

Comment:
This is very light (or completely lacking) error handling

- What happens if we don't have the secret
- If the token is invalid?

Those point to likely a bug on the LMS side or similar. I want a loud sentry exception for now.

Reply:
So I guess LMS will have this secret and use it to decrypt JWEs, so if anyone else ever tries to use the `metadata_jwe` field this decryption will fail?

Again I don't think we should have this encryption/decryption here. I think we've essentially got a bit of "hidden authentication" in the middle of h here. This service is not the right place for authentication to be happening in h. In general authentication shouldn't be happening in the service layer. API requests from LMS-context clients to h are already authenticated in the way described in h's API docs (using a bearer token in an `Authorization` header) and by the time we reach the services layer this authentication has already been verified higher up in h's authentication/security layer.